Client/Helpers/FxALoginHelper.swift

```diff
@@ -0,0 +1,191 @@
+
+        let bookmarkAction = UIMutableUserNotificationAction()
+        bookmarkAction.identifier = SentTabAction.Bookmark.rawValue
+        bookmarkAction.title = NSLocalizedString("Bookmark", comment: "Bookmark a URL - https://bugzilla.mozilla.org/attachment.cgi?id=8624438, https://bug1157303.bugzilla.mozilla.org/attachment.cgi?id=8624440")
+        bookmarkAction.activationMode = UIUserNotificationActivationMode.Foreground
+        bookmarkAction.destructive = false
+        bookmarkAction.authenticationRequired = false
+
+        let readingListAction = UIMutableUserNotificationAction()
+        readingListAction.identifier = SentTabAction.ReadingList.rawValue
+        readingListAction.title = NSLocalizedString("Add to Reading List", comment: "Add URL to the reading list - https://bugzilla.mozilla.org/attachment.cgi?id=8624438, https://bug1157303.bugzilla.mozilla.org/attachment.cgi?id=8624440")
+        readingListAction.activationMode = UIUserNotificationActivationMode.Foreground
+        readingListAction.destructive = false
+        readingListAction.authenticationRequired = false
+
+        let sentTabsCategory = UIMutableUserNotificationCategory()
+        sentTabsCategory.identifier = TabSendCategory
+        sentTabsCategory.setActions([readingListAction, bookmarkAction, viewAction], forContext: UIUserNotificationActionContext.Default)
+
+        sentTabsCategory.setActions([bookmarkAction, viewAction], forContext: UIUserNotificationActionContext.Minimal)
+
+        application?.registerUserNotificationSettings(UIUserNotificationSettings(forTypes: UIUserNotificationType.Alert, categories: [sentTabsCategory]))
+    }
+
+    func userDidRegister(notificationSettings notificationSettings: UIUserNotificationSettings) {
+        // Record that we have asked the user, and they have given an answer.
+        profile?.prefs.setBool(true, forKey: applicationDidRequestUserNotificationPermissionPrefKey)
+
+        guard notificationSettings.types != .None else {
+            return readyForSyncing()
+        }
+
+        if AppConstants.MOZ_FXA_PUSH {
+            application?.registerForRemoteNotifications()
+        } else {
+            readyForSyncing()
+        }
+    }
+
+    func apnsRegisterDidSucceed(apnsToken apnsToken: String) {
+        pushClient.register(apnsToken).upon { res in
+            if let pushRegistration = res.successValue {
+                return self.pushRegistrationDidSucceed(apnsToken: apnsToken, pushRegistration: pushRegistration)
+            }
+            self.apnsRegisterDidFail()
+        }
+    }
+
+    func apnsRegisterDidFail() {
+        readyForSyncing()
+    }
+
+    func pushRegistrationDidSucceed(apnsToken apnsToken: String, pushRegistration: PushRegistration) {
+        account.pushRegistration = pushRegistration
+        readyForSyncing()
+    }
+
+    func pushRegistrationDidFail() {
+        readyForSyncing()
+    }
+
+    func readyForSyncing() {
```

Comment:
Just a random thought:

I noticed that most of the methods in this helper map to various states we're in with regards to registering for push notifications. I wonder if it would be more clear and concise to collapse these into a state machine - similar to what we do for the account process - for transitioning between each stage.

This might also give this class stronger meaning than being named a 'helper'. It could be called `PushRegistrationStateMachine` or something.